src/vendor/libbacktrace_2446c66076480ce07a6bd868badcbceb3eeecc2e_debuginfod_patch.diff

```diff
@@ -0,0 +1,331 @@
++void debuginfod_set_progressfn(debuginfod_client *c,
++			       debuginfod_progressfn_t fn);
++
++void debuginfod_set_verbose_fd(debuginfod_client *c, int fd);
++
++/* Set the user parameter.  */
++void debuginfod_set_user_data (debuginfod_client *client, void *value);
++
++/* Get the user parameter.  */
++void* debuginfod_get_user_data (debuginfod_client *client);
++
++/* Get the current or last active URL, if known.  */
++const char* debuginfod_get_url (debuginfod_client *client);
++
++/* Returns set of x-debuginfod* header lines received from current or
++   last active transfer, \n separated, if known. */
++const char* debuginfod_get_headers(debuginfod_client *client);
++
++/* Add an outgoing HTTP request  "Header: Value".  Copies string.  */
++int debuginfod_add_http_header (debuginfod_client *client, const char* header);
++
++/* Release debuginfod client connection context handle.  */
++void debuginfod_end (debuginfod_client *client);
++
++#ifdef __cplusplus
++}
++#endif
++
++
++#endif /* _DEBUGINFOD_CLIENT_H */
+diff --git a/elf.c b/elf.c
+index fb54165..78efe22 100644
+--- a/elf.c
++++ b/elf.c
+@@ -30,6 +30,8 @@ STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
+ IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ POSSIBILITY OF SUCH DAMAGE.  */
+ 
++#define _GNU_SOURCE
++
+ #include "config.h"
+ 
+ #include <errno.h>
+@@ -38,6 +40,7 @@ POSSIBILITY OF SUCH DAMAGE.  */
+ #include <sys/types.h>
+ #include <sys/stat.h>
+ #include <unistd.h>
++#include <dlfcn.h>
+ 
+ #ifdef HAVE_DL_ITERATE_PHDR
+ #include <link.h>
+@@ -45,6 +48,7 @@ POSSIBILITY OF SUCH DAMAGE.  */
+ 
+ #include "backtrace.h"
+ #include "internal.h"
++#include "debuginfod_support.h"
+ 
+ #ifndef S_ISLNK
+  #ifndef S_IFLNK
+@@ -851,12 +855,86 @@ elf_readlink (struct backtrace_state *state, const char *filename,
+ 
+ #define SYSTEM_BUILD_ID_DIR "/usr/lib/debug/.build-id/"
+ 
++typedef debuginfod_client*(*debuginfod_begin_t)(void);
++
++typedef int(*debuginfod_find_debuginfo_t)(debuginfod_client *client,
++                                            const unsigned char *build_id,
++                                            int build_id_len,
++                                            char **path);
++typedef void(*debuginfod_end_t)(debuginfod_client *client);
++
++ 
++static debuginfod_find_debuginfo_t the_debuginfod_find_debuginfo = NULL;
++static debuginfod_begin_t the_debuginfod_begin = NULL;
++static debuginfod_end_t the_debuginfod_end = NULL;
++static int debuginfod_guard = 0;
++static int debuginfod_dlopen_failed = 0;
++ 
++static int
++elf_open_debugfile_by_debuginfod (const char *buildid_data,
++			       size_t buildid_size,
++			       backtrace_error_callback error_callback,
++			       void *data)
++{
++  int ret;
++  int does_not_exist;
++
++  if (debuginfod_dlopen_failed) {
++      return -1;
++  }
++
++  if (the_debuginfod_find_debuginfo == NULL) {
```

Comment:
Defensive programming nit: check for the last one to be assigned, instead of the first, so that if we enter this function with `the_debuginfod_find_debuginfo` set (because `dlopen` and `dlsym("debuginfod_find_debuginfo")` succeed but `dlsym("debuginfod_end")` fails, we don't skip this block and try calling a null pointer.
```suggestion
+  if (the_debuginfod_end == NULL) {
```

Or alternatively, set any that have been assigned to `nullptr` on error. That's perhaps more readable, though this suggestion is more succinct.